File: tl/src/main/java/com/github/badoualy/telegram/tl/core/TLGzipObject.java
package com.github.badoualy.telegram.tl.core;

import com.github.badoualy.telegram.tl.StreamUtils;
import com.github.badoualy.telegram.tl.TLContext;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
public class TLGzipObject extends TLObject {

    public static final int CONSTRUCTOR_ID = 0x3072cfa1;

    private byte[] packedData;

    public TLGzipObject() {

    }

    public TLGzipObject(byte[] packedData) {
        this.packedData = packedData;
    }


    @Override
    public void serializeBody(OutputStream stream) throws IOException {
        StreamUtils.writeTLBytes(packedData, stream);
    }

    @Override
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {
        packedData = StreamUtils.readTLBytes(stream);
    }

    @Override
    public String toString() {
        return "gzip_packed#3072cfa1";
    }

    @Override
    public int getConstructorId() {
        return CONSTRUCTOR_ID;
    }

    public byte[] getPackedData() {
        return packedData;
    }

    public void setPackedData(byte[] packedData) {
        this.packedData = packedData;
    }
}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/core/TLGzipObject.kt
package com.github.badoualy.telegram.tl.core

import com.github.badoualy.telegram.tl.StreamUtils
import com.github.badoualy.telegram.tl.TLContext

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.Arrays

import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLBytesSerializedSize

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see [http://github.com/badoualy/kotlogram](http://github.com/badoualy/kotlogram)
 */
class TLGzipObject(var packedData: ByteArray? = null) : TLObject() {

    override val constructorId: Int
        get() = CONSTRUCTOR_ID

    @Throws(IOException::class)
    override fun serializeBody(stream: OutputStream) {
        StreamUtils.writeTLBytes(packedData, stream)
    }

    @Throws(IOException::class)
    override fun deserializeBody(stream: InputStream, context: TLContext) {
        packedData = StreamUtils.readTLBytes(stream)
    }

    override fun computeSerializedSize(): Int =
            CONSTRUCTOR_ID + computeTLBytesSerializedSize(packedData!!.size)

    override fun toString() = "gzip_packed#3072cfa1"

    override fun equals(other: Any?): Boolean {
        if (other !is TLGzipObject)
            return false
        if (this === other)
            return true

        return Arrays.equals(packedData, other.packedData)
    }

    override fun hashCode() = packedData?.let { Arrays.hashCode(it) } ?: 0

    companion object {
        const val CONSTRUCTOR_ID = 0x3072cfa1
    }
}
